pkg/kubelet/container/cache.go

```diff
@@ -142,6 +158,28 @@ func (c *cache) get(id types.UID) *data {
 // Otherwise, it returns nil. The caller should acquire the lock.
 func (c *cache) getIfNewerThan(id types.UID, minTime time.Time) *data {
 	d, ok := c.pods[id]
+	if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
+		// Evented PLEG has CREATED, STARTED, STOPPED and DELETED events
+		// However if the container creation fails for some reason there is no
+		// CRI event received by the kubelet and that pod will get stuck a
+		// GetNewerThan call in the pod workers. This is reproducible with
+		// the node e2e test, https://bit.ly/3yJur4P
+		// which forces failure during pod creation. This issue also exists in
+		// Generic PLEG but since it updates global timestamp periodically
+		// the GetNewerThan call gets unstuck.
+
+		// During node e2e tests, it was observed this change does not have any
+		// adverse impact on the behaviour of the Generic PLEG as well.
+		switch {
+		case !ok:
+			return makeDefaultData(id)
```

Comment:
so the difference when EventedPLEG is enabled is that:

when EventedPLEG is enabled : when pod ID is not found in `c.pods`, getIfNewerThan always returns default data

when EventedPLEG is disabled : when pod ID is not found in `c.pods`, getIfNewerThan only return default data **if the global timestamp is newer**


Why the difference? I see that evented.go updates the global timestamp every 5s. Is it too infrequent and therefore pod workers might get stuck for 5 seconds?

Reply:
Actually the change works fine even when the Evented PLEG is not enabled. I didn't observe any different behaviour in the tests. However, I thought it would be safer to keep the existing behaviour as is, and change the behavior only when Evented PLEG is enabled just out of abundance of caution. We could allow this change to go through some rigorous testing before making it default without gating it with Evented PLEG.

Reply:
yes I understand that we want to keep the existing behavior as is. But I wanted to understand why eventedPLEG always return default data when pod ID is not found, as opposed to the behavior of existing pleg.

Reply:
While using Evented PLEG there can be a case, where CRI fails to create container and the podID would be unknown in the cache, which is similar to node e2e test https://bit.ly/3yJur4P, essentially we noticed that without this change the pod workers get stuck and would cause the test to fail. The default data returned from this function `getIfNewerThan` remains stuck. This change helps it to get unblocked when no such pod ID exists in the cache without it having to depend on the global timestamp. 

Noticeably, changing this behavior also doesn't have any adverse impact while just using Generic PLEG, however as @harche suggested we could allow this change to go through some rigorous testing before putting this change out of the feature gate.

Reply:
Just to clarify, do you mean if we change this to the following:

```
if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
		switch {
		case !ok&& globalTimestampIsNewer:
			return makeDefaultData(id)
...
``` 

, (in other words, same as generic pleg (https://github.com/kubernetes/kubernetes/pull/111384/files#diff-b010d1c9ff51c8975875be97aeebd3c8c6206128f520115752f9dc007e8afcd1R187), then `getIfNewerThan` will get stuck when container creation failure happens?

generic PLEG mitigates this problem by periodically updating the global timestamp, which I understand.  But doesn't evented PLEG update global timestamp every 5 seconds too (https://github.com/kubernetes/kubernetes/pull/111384/files#diff-465aae123f6877816ff0428968fc74f503fe15b27b434480bf57b79e62c7fa2fR114)? So `getIfNewerThan` should not stuck for more than 5 seconds, right?

This change looks ok to me because in theory the cache entry is always up-to-date with evented pleg, just trying to understand why the issue happens.

Reply:
Yes, that correctly describes the scenario.

Reply:
I had the same question when I was reviewing this, I was also trying to understand the reason for the difference here since it is not immediately clear :). Can we maybe summarize the reasoning for this change as a comment in the code?

Reply:
We tried to capture this in the [comment](https://github.com/kubernetes/kubernetes/pull/111384/files#diff-b010d1c9ff51c8975875be97aeebd3c8c6206128f520115752f9dc007e8afcd1R161) above the code, but feel free to suggest any amendments.

Reply:
thanks for the comment, i was similar to @bobbypage when catching up on latest.